monadic thread macros

blendmaster opened the pull request:
Probably easiest to think about as a generalization of `some->` to all monads.

They're kinda cool, but I honestly had a harder time finding uses than actually writing the code. The amount of monadic functions that need composing _and_ have extra pure arguments might not be rich enough to justify more macros.

Note that `>>=` is already variadic, and even has a `m->` compatible example:

```clj
(>>= (just 1) (comp just inc) (comp just inc))
```

If these threading macros do end up being desirable, an alternative name for `m->`/`m->>` would be to pun `->=` and `->>=` in a mix of haskell and clojure convention.

yurrriq commented:
Cool! I like `->=` and `->>=`. Down the road, I could see great utility in [`->=<>`][1] et al.

[1]: https://github.com/rplevy/swiss-arrows/blob/bd78101b43db8c81e7679b4ea96e9a2b972529be/src/swiss/arrows.clj#L34-L41

blendmaster commented:
Can you think of a useful example domain for this, both to vet its usefulness and to write tests around it? I'm still having trouble coming up with something that isn't trivial.

>Down the road, I could see great utility in `->=<>`, et al.

I can add a generalization of [`as->`][0] from clojure core; That covers the bases without being too abstruse. There's also [`cond->`][1] I guess, but that's weird enough even without monads.

[0]: https://clojuredocs.org/clojure.core/as-%3E
[1]: https://clojuredocs.org/clojure.core/cond-%3E

yurrriq commented:
> There's also cond-> I guess, but that's weird enough even without monads.

:laughing: True. I rarely use `as->`, to be honest.

> Can you think of a useful example domain for this...?

I'll try to think of some and get back to you.

Re: `->>=<>` et al. I meant specifically the `<>` placeholder capabilities, as found in [swiss-arrows][].

Something like `(->>=<> mv mf-1 (mf-2 x <> y) (mf-3 a <> b c))` and so forth.

In production code, I've often found the need to create anonymous functions or do otherwise unnecessary currying to keep `(>>= ...)` forms succinct. With these threading macros (and maybe the diamond-wand-style ones), it would be significantly easier in those cases to write short, sweet, effective code.

e.g. `(>>= mv (fn [x] (f x y)))` or currying `f` for the sole purpose of enabling `(>>= mv (f y))`, which is a bit too opaque for my taste.

[swiss-arrows]: https://github.com/rplevy/swiss-arrows

niwinz commented:
Nice work with `m->` and `m->>` ;)

+1 to add also something like`>>=<>` :P

dialelo commented:
Cool macros, seem like a great candidate for the `cats.labs.sugar` namespace mentioned in #122. Note that macros must be written in Clojure and hence have to be defined using a reader conditional as seen in `cats.core`.

Since we can experiment however we like under `cats.labs.sugar` feel free to add more arrow macros that you think will be useful!

blendmaster commented:
Reamed to `->=` and `->>=`, moved to `cats.labs.sugar` namespace, and added some (admittedly weak) tests. Suggestions on what else to check welcome. as are any doc improvements (I wrote the macros, so it's hard to know what questions people will have).

Also added `as->=`, which should be exactly as it sounds (implementation is literally `clojure.core/as->` copy pasted, but with `mlet` instead of `let`. It's slightly more typing than then hypothetical `->=<>`, but avoids the name collision with Monoid `<>`. 

One point of interest is that `as->` _requires_ the binding to be present in every threaded form; there's no implicit behavior for a single-symbol form or forms without the binding present, like `->=<>` has. We _could_ implement the same, and have `as->=`/`as->>=` for the appropriate default behavior, but then it'd be different than core's `as->`. Also, I feel like the implicit behavior is uncommon enough to be "too sugary"; this is probably the same rationale for `as->` in core instead of diamond wands.

yurrriq commented:
Wow, I never made the connection between `as->` and `-<>>` before. The main difference between the the wands and `as->` is that they default the their vanilla arrow equivalents when `<>` is not present in a form.

It'd be cool to get monadic furculas going too. Especially for Either.

yurrriq commented:
It's d-day at work but I should have some free time to play around, once I submit to the App Store.

blendmaster commented:
> In practice, I find it really handy when most of my threading conforms to -> or ->>` but only one or two need the result to be threaded in a different place. I feel it's a handy tool to have in the belt, even if it's not used often.

FWIW, `as->` fits fairly nicely within a `->` form:

```clj
(-> a
  b
  (c d)
  (as-> $
    (e f $ g))
  h)
```

More annoying in a `->>` form though, you pretty much have to wrap in `->`:

```clj
(-> a
  (->>
    b
    (c d))
  (as-> $
    (e f $ g))
  (->> h))
```

I guess you could make `as->>` with a flipped last argument for that, but still a bit gross.

>It'd be cool to get monadic furculas going too. Especially for Either.

I've stared at the furcula section of swiss-arrows, but I still don't quite get how they're useful,
or how they'd translate into a monadic context. It seems like it basically desugars to:

```clj
(macroexpand-1 `(-< 1 (+ 1) (+ 2) (+ 3)))
=>
(let [$ 1]
  [(+ $ 1) (+ $ 2) (+ $ 3)])
``` 

That straightforward conversion using `mlet` instead of `let` doesn't seem particularly useful, and could be emulated with `ap #(<- % ...) monadic-thing)`. I guess there's something in that it's kind of a non-deterministic computation, a.k.a list monad, but then it's still not clear how that generalizes. lisp syntax is lists, there's not an async syntax or a maybe syntax for the furcula to sugar over.

yurrriq commented:
> FWIW, as-> fits fairly nicely within a -> form:

I like and can handle that.

> More annoying in a ->> form though, you pretty much have to wrap in ->:

True.

I'm not overly committed to getting the "wand" macros into cats. With `->=`, `->>=` and `->as=` you can accomplish most of what the wands would give us.


Re: furcula. I just re-read their definitions and that's not what I meant nor how I remembered them, oops.

Rather, what I thought would be cool is some sugar for chaining pairs of forms, one for an "unsuccessful" bind (`Nothing`, `Left`, `Failure`, etc) and one for a "successful bind" (`Just`, `Right`, `Success`, etc). 

Quick thought dump (will try to clarify some other time):

```clojure
(require '[cats.core :refer [mlet]]
         '[cats.monad.either :as either])

(= (-< either-value
       [(f-left x y) (f-right x y)]
       (g-right z))
   (let [step1 (either/branch either-value
                 #(f-left % x y)
                 #(f-right % x y))
         step2 (mlet [v step1]
                 (g-right v z))]
     step2))
```

This means we'd have to generalize `either/branch`, but I think that'd be pretty simple. Something along the lines of:

```clojure
(require '[cats.core :refer [>>=]]
         '[cats.monad.either :as either]
         '[cats.monad.maybe :as maybe])

(defn mbranch [mv f g]
  (if (= mv (>>= mv identity))
    (f (extract mv))
    (>>= mv g)))

(defn maybe-example [mv]
  (mbranch mv
    (fn [_nothing] "Oh no!")
    (fn [value]    (str "You've got " value))))

(maybe-example (maybe/just "something"))
;;=> "You've got something"

(maybe-example (maybe/nothing))
;;=> "Oh no!"

(defn either-example [ev]
  (mbranch ev
    (fn [lv] (str "Do something with left: "  lv))
    (fn [rv] (str "Do something with right: " rv))))

(either-example (either/left 0))
;;=> "Do something with left: 0"

(either-example (either/right 42))
;;=> "Do something with right: 42"
```

Then again, maybe that's too much sugar :laughing:

blendmaster commented:
>Rather, what I thought would be cool is some sugar for chaining pairs of forms, one for an "unsuccessful" bind (Nothing, Left, Failure, etc) and one for a "successful bind" (Just, Right, Success, etc). 

Hmm, could be interesting. I'm not sure it really generalizes to all monads though. Maybe/Try/Either all have the same "shape", but what are the branches on, say, a core.async channel? Also,  `either/branch` is already a nice pure fold/catamorphism/visitor, so you can use regular `->` and `#()` reader macro without much overhead:

```clj
(-> either-val
  (branch #(f-left % x y) #(f-right % x y))
  (fmap #(g-right % z)))

;; or more symmetically
(-> either-val
  (branch #(f-left % x y) #(f-right % x y))
  (branch id              #(g-right % z)))
```

You can macro away the `branch` and `%` repetition as you please.

`branch` is also the `bimap` for Either's Bifunctor instance; I don't know enough about them (and its friends Biapplicative and Bimonad) know if that's useful here though.

https://www.fpcomplete.com/school/to-infinity-and-beyond/pick-of-the-week/profunctors#bifunctors
https://stackoverflow.com/questions/13556314/biapplicative-and-bimonad

yurrriq commented:
@blendmaster, agreed.

:+1: For messing with Bifunctors (and Biapplicatives) in cats.

yurrriq commented:
FYI, I [spec'd out a `Bifunctor` protocol][1] and [made `either/context` satisfy it][2], and

[1]: https://github.com/funcool/cats/commit/1db6c64980a75e20dde86816248c29b5b0ab2dd2
[2]: https://github.com/funcool/cats/commit/0bcb3bf241d339bf2cbf2693d9ea04d4a07ecf90

> branch is also the bimap for Either's Bifunctor instance

is not the case.
```clojure
(p/-bimap context #(+ % 2) #(- % 2) (left 2))
;;=> #<Left 4>
(branch (left 2) #(+ % 2) #(- % 2))
;;=> 4

(p/-bimap context #(+ % 2) #(- % 2) (right 2))
;;=> #<Right 0>
(branch (right 2) #(+ % 2) #(- % 2))
0
```


The different is interesting, and also...

```clojure
(= (branch (left 2) (comp left #(+ % 2)) (comp right #(- % 2)))
   (p/-bimap context #(+ % 2) #(- % 2) (left 2)))

(= (branch (left 2) #(+ % 2) #(- % 2))
   @(p/-bimap context #(+ % 2) #(- % 2) (left 2))
   (p/-extract (p/-bimap context #(+ % 2) #(- % 2) (left 2))))
```

As such, I've altered the definitions of branch{-left,-right,} accordingly [here][3].
[3]: https://github.com/funcool/cats/commit/eb826c103b9233653f9d08a7316c5a9395a915d9

dialelo commented:
> FYI, I spec'd out a Bifunctor protocol and made either/context satisfy it

This would be so handy for promises and my ongoing work on funcool/muse#5, please open a pull request!

> branch is also the bimap for Either's Bifunctor instance
> is not the case.

As @yurrriq points out its not, it extracts the value from the either which may not be available in similar "shaped" types like promises. Looking forward to add `Bifunctor`to cats.

blendmaster commented:
Ah, I thought that `branch` kept things in the Either functor. In any case, I'm happy enough with `->=` et al.

blendmaster commented:
rebased so it'll merge cleanly.

dialelo commented:
Great, thanks a lot!

yurrriq commented:
:tada: Thanks, @blendmaster! I'm loving the discussions of late.
